File: sdk-typescript/test/canon.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { canonicalize, canonicalHash } from "../src/canon.ts";

// Shared corpus lives at repo-root /tests; resolve relative to this file so the
// test is independent of the working directory.
const TESTS_DIR = join(import.meta.dirname, "..", "..", "tests");

function loadCorpus(name: string): any {
  return JSON.parse(readFileSync(join(TESTS_DIR, name), "utf8"));
}

// Cross-language contract: the TS canonicalizer MUST reproduce every
// `canonical` string byte-for-byte (this is what guarantees action_hash parity
// with the Go SDK, the Python SDK, and the Rust gateway).
test("canonical action vectors match the shared corpus byte-for-byte", () => {
  const corpus = loadCorpus("canonical_action_vectors.json");
  assert.ok(Array.isArray(corpus.vectors) && corpus.vectors.length > 0, "no vectors");
  for (const vec of corpus.vectors) {
    assert.equal(canonicalize(vec.tool_call), vec.canonical, `vector ${vec.name}`);
  }
});

// End-to-end canonicalizer + SHA-256: for each shared receipt vector,
// SHA-256(canonicalize(body)) — body is every field except receipt_hash — MUST
// equal the pinned receipt_hash.
test("receipt chain vectors reproduce pinned receipt_hash values", () => {
  const corpus = loadCorpus("receipt_chain_vectors.json");
  assert.ok(Array.isArray(corpus.receipts) && corpus.receipts.length > 0, "no receipts");
  for (const rec of corpus.receipts) {
    const { receipt_hash, ...body } = rec;
    assert.equal(canonicalHash(body), receipt_hash, `receipt ${rec.event_id}`);
  }
});

// Fail-closed number rule.
test("non-finite numbers are rejected", () => {
  assert.throws(() => canonicalize({ x: Infinity }));
  assert.throws(() => canonicalize({ x: NaN }));
});

File: sdk-go/canon/canon_test.go
package canon

import (
	"encoding/json"
	"os"
	"testing"
)

// loadCorpus decodes a shared corpus file with UseNumber() so JSON numbers are
// preserved as json.Number (not coerced to float64) — exactly how the SDK must
// treat tool-call/receipt numbers to stay byte-identical with Python/Rust.
func loadCorpus(t *testing.T, path string) map[string]any {
	t.Helper()
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer f.Close()
	dec := json.NewDecoder(f)
	dec.UseNumber()
	var m map[string]any
	if err := dec.Decode(&m); err != nil {
		t.Fatalf("decode %s: %v", path, err)
	}
	return m
}

// TestCanonicalActionVectors is the cross-language contract: the Go
// canonicalizer MUST reproduce every `canonical` string in the shared corpus
// byte-for-byte (this is what guarantees action_hash parity with the gateway).
func TestCanonicalActionVectors(t *testing.T) {
	data := loadCorpus(t, "../../tests/canonical_action_vectors.json")
	vectors, ok := data["vectors"].([]any)
	if !ok {
		t.Fatal("corpus has no `vectors` array")
	}
	if len(vectors) == 0 {
		t.Fatal("corpus has zero vectors")
	}
	for _, v := range vectors {
		vec := v.(map[string]any)
		name, _ := vec["name"].(string)
		want, _ := vec["canonical"].(string)
		got, err := Canonicalize(vec["tool_call"])
		if err != nil {
			t.Fatalf("%s: Canonicalize: %v", name, err)
		}
		if got != want {
			t.Errorf("%s: canonical mismatch\n got:  %s\n want: %s", name, got, want)
		}
	}
}

// TestReceiptChainVectors validates the canonicalizer + SHA-256 end-to-end: for
// each shared receipt vector, SHA-256(canonicalize(body)) — where body is every
// field except `receipt_hash` — MUST equal the pinned `receipt_hash`.
func TestReceiptChainVectors(t *testing.T) {
	data := loadCorpus(t, "../../tests/receipt_chain_vectors.json")
	receipts, ok := data["receipts"].([]any)
	if !ok {
		t.Fatal("corpus has no `receipts` array")
	}
	if len(receipts) == 0 {
		t.Fatal("corpus has zero receipts")
	}
	for _, r := range receipts {
		rec := r.(map[string]any)
		want, _ := rec["receipt_hash"].(string)
		body := make(map[string]any, len(rec))
		for k, val := range rec {
			if k == "receipt_hash" {
				continue
			}
			body[k] = val
		}
		got, err := CanonicalHash(body)
		if err != nil {
			t.Fatalf("receipt %v: CanonicalHash: %v", rec["event_id"], err)
		}
		if got != want {
			t.Errorf("receipt %v: hash mismatch\n got:  %s\n want: %s", rec["event_id"], got, want)
		}
	}
}

// TestNonFiniteFloatRejected confirms the fail-closed number rule.
func TestNonFiniteFloatRejected(t *testing.T) {
	if _, err := Canonicalize(map[string]any{"x": math_Inf()}); err == nil {
		t.Error("expected error for non-finite float, got nil")
	}
}

func math_Inf() float64 { return 1.0 / zero() }
func zero() float64     { return 0.0 }
